File: ext/RmlUi.Native/src/FileInterface.cpp
#include "RmlNative.h"
#include "RmlUi/Core/FileInterface.h"
#include <iostream>
#include <string>
#include <cstdlib>

typedef Rml::FileHandle(*onOpen)(const char* path);
typedef void(*onClose)(Rml::FileHandle file);
typedef const char* (*onLoadFile)(const char* path);
typedef size_t(*onRead)(uint8_t* buffer, size_t size, Rml::FileHandle file);
typedef bool(*onSeek)(Rml::FileHandle file, long offset, int origin);
typedef size_t(*onTell)(Rml::FileHandle file);
typedef size_t(*onLength)(Rml::FileHandle file);

class FileInterface : public Rml::FileInterface {
private:
	::onOpen m_onOpen;
	::onClose m_onClose;
	::onLoadFile m_onLoadFile;
	::onRead m_onRead;
	::onSeek m_onSeek;
	::onTell m_onTell;
	::onLength m_onLength;

public:
	explicit FileInterface(
		::onOpen onOpen,
		::onClose onClose,
		::onLoadFile onLoadFile,
		::onRead onRead,
		::onSeek onSeek,
		::onTell onTell,
		::onLength onLength
	) {
		m_onOpen = onOpen;
		m_onClose = onClose;
		m_onLoadFile = onLoadFile;
		m_onRead = onRead;
		m_onSeek = onSeek;
		m_onTell = onTell;
		m_onLength = onLength;
	}

	Rml::FileHandle Open(const Rml::String& path) override {
		return (*m_onOpen)(path.c_str());
	}

	void Close(Rml::FileHandle file) override {
		(*m_onClose)(file);
	}

	size_t Read(void* buffer, size_t size, Rml::FileHandle file) override {
		uint8_t* bytes = static_cast<uint8_t*>(std::malloc(sizeof(uint8_t) * size));
		size_t length = (*m_onRead)(bytes, size, file);

		// shut up I know
		memcpy(buffer, bytes, length);

		return length;
	}

	bool Seek(Rml::FileHandle file, long offset, int origin) override {
		return (*m_onSeek)(file, offset, origin);
	}

	size_t Tell(Rml::FileHandle file) override {
		return (*m_onTell)(file);
	}

	size_t Length(Rml::FileHandle file) override {
		return (*m_onLength)(file);
	}

	bool LoadFile(const Rml::String& path, Rml::String& out_data) override {
		out_data = (*m_onLoadFile)(path.c_str());
		return (out_data != "");
	}
};

RMLUI_CAPI void* rml_FileInterface_New(
	::onOpen onOpen,
	::onClose onClose,
	::onLoadFile onLoadFile,
	::onRead onRead,
	::onSeek onSeek,
	::onTell onTell,
	::onLength onLength
) {
	return new FileInterface(
		onOpen,
		onClose,
		onLoadFile,
		onRead,
		onSeek,
		onTell,
		onLength
	);
}

File: ext/RmlUI.Net/FileInterface.cs
using System;

namespace RmlUiNet
{
    public abstract class FileInterface : RmlBase<FileInterface>
    {
        private Native.FileInterface.OnOpen _onOpen;
        private Native.FileInterface.OnClose _onClose;
        private Native.FileInterface.OnRead _onRead;
        private Native.FileInterface.OnSeek _onSeek;
        private Native.FileInterface.OnTell _onTell;
        private Native.FileInterface.OnLength _onLength;
        private Native.FileInterface.OnLoadFile _onLoadFile;

        public unsafe FileInterface() : base(IntPtr.Zero)
        {
            _onOpen = Open;
            _onClose = Close;
            _onRead = _Read;
            _onSeek = Seek;
            _onTell = Tell;
            _onLength = Length;
            _onLoadFile = LoadFile;

            NativePtr = Native.FileInterface.Create(
                _onOpen,
                _onClose,
                _onLoadFile,
                _onRead,
                _onSeek,
                _onTell,
                _onLength
            );

            ManuallyRegisterCache(NativePtr, this);
        }

        internal unsafe ulong _Read(byte* buffer, ulong size, ulong file)
        {
            byte[] bytes = new byte[size];
            var len = Read(bytes, size, file);

            for (var i = 0; i < (int)len; i++)
            {
                buffer[i] = bytes[i];
            }

            return len;
        }

        public abstract ulong Open(string path);
        public abstract void Close(ulong file);
        public abstract ulong Read(byte[] buffer, ulong size, ulong file);
        public abstract bool Seek(ulong file, long offset, int origin);
        public abstract ulong Tell(ulong file);
        public abstract ulong Length(ulong file);
        public abstract string LoadFile(string path);
    }
}
